### DISCUSSION TITLE
How to return internationalized json response in quarkus graphql when exception

### DISCUSSION THREAD

**Discussion by pipinet** [quarkusio/quarkus#40982] [Q&A]
How to return internationalized json response in quarkus graphql when exception

[a comment by an automated account is not shown]

**Comment by bestbreadmakerst** [comment]
I’ve worked with Quarkus GraphQL and can help with this. First, create a custom exception. This will handle the internationalized messages. Use the @GraphQLException annotation. It allows you to catch and process exceptions in GraphQL.

 @GraphQLException
public class CustomGraphQLException extends RuntimeException {
    private final String messageKey;

    public CustomGraphQLException(String messageKey) {
        this.messageKey = messageKey;
    }

    public String getMessageKey() {
        return messageKey;
    }
}

Next, create a message resource bundle. This will store your internationalized messages. Place it in the src/main/resources folder.

messages_en.properties
messages_fr.properties

Then, create a service to handle localization. Inject this service into your exception handler.

@ApplicationScoped
public class MessageService {
    @Inject
    private Locale locale;

    public String getMessage(String key) {
        ResourceBundle bundle = ResourceBundle.getBundle("messages", locale);
        return bundle.getString(key);
    }
}
Finally, implement the exception handler. Use the service to get the localized message.
public class GraphQLExceptionHandler implements GraphQLErrorHandler {
    @Inject
    MessageService messageService;

    @Override
    public List<GraphQLError> processErrors(List<GraphQLError> errors) {
        return errors.stream()
                .map(error -> {
                    if (error.getException() instanceof CustomGraphQLException) {
                        CustomGraphQLException ex = (CustomGraphQLException) error.getException();
                        String localizedMessage = messageService.getMessage(ex.getMessageKey());
                        return GraphqlErrorBuilder.newError().message(localizedMessage).build();
                    }
                    return error;
                })
                .collect(Collectors.toList());
    }
}
Configure the handler in application.properties.

quarkus.graphql.error-handler=com.example.GraphQLExceptionHandler

This setup ensures that your GraphQL exceptions return internationalized JSON responses. For more details, check the Quarkus documentation and the GraphQL Java documentation.

Good luck with your implementation!

**Reply by pipinet** [reply]
@bestbreadmakerst thanks for your reply.  exception will be thrown by typed  message in my project. so exception message will internationalized by the current request context locale when created. is there any way make SomeMessages internationalized?

```java
@MessageBundle(projectCode = "Q")
public interface SomeMessages {
    SomeMessages INSTANCE = getBundle(SomeMessages.class);
    @Message("widget template id:{0} is not existed", format = Message.Format.MESSAGE_FORMAT)
    NotFoundException templateNotFound(String templateId);
}


class SomeService {
   public findById(String id){
      if(not found){
          throw  SomeMessages.INSTANCE.templateNotFound(id);
      }
  }
}

```

**Reply by fabrii** [reply]
Hi! Any hint on how to do this with graphql smallrye?
